File: python/unique_paths_ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        last_square = (len(obstacleGrid) - 1, len(obstacleGrid[0]) - 1)
        if obstacleGrid[last_square[0]][last_square[1]] != 0 or obstacleGrid[0][0] != 0:
            return 0
        memo = { last_square: 1 }
        def backtrack(location: Tuple[int, int]) -> int:
            if location not in memo:
                subProblems = []
                x, y = location
                if x + 1 < len(obstacleGrid) and obstacleGrid[x + 1][y] == 0:
                    subProblems.append(backtrack((x + 1, y)))
                if y + 1 < len(obstacleGrid[0]) and obstacleGrid[x][y + 1] == 0:
                    subProblems.append(backtrack((x, y + 1)))
                memo[location] = sum(subProblems)
            return memo[location]
        return backtrack((0, 0))

File: java/uniquePathsII.java
class Solution {
    
    Map<Integer, Map<Integer, Integer>> memo = new HashMap();
    int n = 0;
    int m = 0;
    int[][] obstacleGrid = new int[0][0];
    int curX = 0;
    int curY = 0;
    
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        n = obstacleGrid.length;
        m = obstacleGrid[0].length;
        if (obstacleGrid[n - 1][m - 1] != 0 || obstacleGrid[0][0] != 0) {
            return 0;
        }
        this.obstacleGrid = obstacleGrid;
        Map<Integer, Integer> lastKey = new HashMap();
        lastKey.put(m - 1, 1);
        memo.put(n - 1, lastKey);
        return backtrack();
    }
    
    private int backtrack() {
        if (!memo.keySet().contains(curX) || !memo.get(curX).keySet().contains(curY)) {
            curX++;
            int leftSubproblem = curX < n && obstacleGrid[curX][curY] == 0 ? backtrack() : 0;
            curX--;
            curY++;
            int rightSubproblem = curY < m && obstacleGrid[curX][curY] == 0 ? backtrack() : 0;
            curY--;
            int result = leftSubproblem + rightSubproblem;
            if (memo.keySet().contains(curX)) {
                memo.get(curX).put(curY, result);
            } else {
                Map<Integer, Integer> curKey = new HashMap();
                curKey.put(curY, result);
                memo.put(curX, curKey);
            }
        }
        return memo.get(curX).get(curY);
    }
}